File: dotnet/Allmark/Inline/LinkRule.cs
namespace Allmark.Inline;

using Allmark.Types;

public static class LinkRule
{
    public static InlineRule Create()
    {
        return new InlineRule
        {
            Name = "link",
            Test = TestLink,
            Precedence = 15,
        };
    }

    private static bool TestLink(InlineParserState state, MarkdownNode parent)
    {
        var ch = Utils.GetChar(state.Src, state.I);

        if (!state.IsEscaped)
        {
            if (ch == '[')
            {
                return TestLinkOpen(state, parent);
            }

            if (ch == '!' && state.I + 1 < state.Src.Length && state.Src[state.I + 1] == '[')
            {
                return TestImageOpen(state, parent);
            }

            if (ch == ']')
            {
                return TestLinkClose(state, parent);
            }
        }

        return false;
    }

    private static bool TestLinkOpen(InlineParserState state, MarkdownNode parent)
    {
        var start = state.I;
        var markup = "[";

        // Add a new text node which may turn into a link
        var text = Utils.NewText(state.ParentIndex + start, state.Line, markup, 0);
        parent.Children!.Add(text);

        state.I++;
        state.Delimiters.Add(new Delimiter { Markup = markup, Start = start, Length = 1, Precedence = Create().Precedence!.Value });

        return true;
    }

    private static bool TestImageOpen(InlineParserState state, MarkdownNode parent)
    {
        var start = state.I;
        var markup = "![";

        // Add a new text node which may turn into an image
        var text = Utils.NewText(state.ParentIndex + start, state.Line, markup, 0);
        parent.Children!.Add(text);

        state.I += markup.Length;
        state.Delimiters.Add(new Delimiter { Markup = markup, Start = start, Length = 1, Precedence = Create().Precedence!.Value });

        return true;
    }

    private static bool TestLinkClose(InlineParserState state, MarkdownNode parent)
    {
        // Loop backwards through delimiters to find a matching one that does
        // not take precedence
        Delimiter? startDelimiter = null;
        var i = state.Delimiters.Count;
        while (i-- > 0)
        {
            var prevDelimiter = state.Delimiters[i];
            if (!prevDelimiter.Handled)
            {
                if (prevDelimiter.Markup == "[" || prevDelimiter.Markup == "![")
                {
                    startDelimiter = prevDelimiter;
                    break;
                }
                else if ((prevDelimiter.Precedence ?? 0) <= Create().Precedence!.Value)
                {
                    // Same or lower precedence delimiters can be skipped over
                    continue;
                }
                else
                {
                    // Higher precedence delimiters block
                    break;
                }
            }
        }

        if (startDelimiter != null)
        {
            // Convert the text node into a link node with a new text child
            // followed by the other children of the parent (if any)
            var j = parent.Children?.Count ?? 0;
            while (j-- > 0)
            {
                var lastNode = parent.Children?[j];
                if (lastNode?.Index == state.ParentIndex + startDelimiter.Start)
                {
                    var start = state.I + 1;
                    var label = state.Src.Substring(
                        startDelimiter.Start + startDelimiter.Markup.Length,
                        state.I - (startDelimiter.Start + startDelimiter.Markup.Length));

                    // "The link text may contain balanced brackets, but not
                    // unbalanced ones, unless they are escaped"
                    var level = 0;
                    for (var k = 0; k < label.Length; k++)
                    {
                        if (label[k] == '\\')
                        {
                            k++;
                        }
                        else if (label[k] == '[')
                        {
                            level++;
                        }
                        else if (label[k] == ']')
                        {
                            level--;
                        }
                    }
                    if (level != 0)
                    {
                        return false;
                    }

                    var isLink = startDelimiter.Markup == "[";

                    var hasInfo = state.I + 1 < state.Src.Length && state.Src[state.I + 1] == '(';
                    var hasRef = state.I + 1 < state.Src.Length && state.Src[state.I + 1] == '[';

                    // "Full and compact references take precedence over shortcut references"
                    // "Inline links also take precedence"
                    LinkReference? link = null;
                    if (hasInfo)
                    {
                        start++;
                        link = Utils.ParseLinkInline(state, start);
                    }
                    else if (hasRef)
                    {
                        start++;
                        for (var k = start; k < state.Src.Length; k++)
                        {
                            if (state.Src[k] == ']')
                            {
                                // Lookup using the text between the [], or if there
                                // is no text, use the label
                                label = k - start > 0 ? state.Src.Substring(start, k - start) : label;
                                label = Utils.NormalizeLabel(label);
                                if (state.Refs.TryGetValue(label, out link))
                                {
                                    state.I = k + 1;
                                }
                                break;
                            }
                        }
                    }

                    if (link == null)
                    {
                        label = Utils.NormalizeLabel(label);
                        state.Refs.TryGetValue(label, out link);
                        if (link != null)
                        {
                            state.I++;
                        }
                    }

                    if (link != null)
                    {
                        var content = lastNode.Content.Substring(startDelimiter.Markup.Length) ?? "";
                        var text = Utils.NewText(lastNode.Index, lastNode.Line, content, 0);
                        text.Length = text.Content.Length;

                        lastNode.Type = isLink ? "link" : "image";
                        lastNode.Info = link.Url;
                        lastNode.Title = link.Title;
                        lastNode.Length = state.ParentIndex + state.I - lastNode.Index;
                        var movedNodes = parent.Children!.Skip(j + 1).ToList() ?? [];
                        parent.Children!.RemoveRange(j + 1, movedNodes.Count);
                        lastNode.Children = [text, .. movedNodes];

                        // "[L]inks may not contain other links, at any level of nesting"
                        if (isLink)
                        {
                            // Remove all the opening delimiters so they won't be picked up in future
                            var d = state.Delimiters.Count;
                            while (d-- > 0)
                            {
                                var prevDelimiter = state.Delimiters[d];
                                if (prevDelimiter.Markup == "[" || prevDelimiter.Markup == "]")
                                {
                                    prevDelimiter.Handled = true;
                                }
                            }
                        }

                        startDelimiter.Handled = true;
                        return true;
                    }

                    startDelimiter.Handled = true;
                    break;
                }
            }
        }

        return false;
    }
}

File: web/src/inline/linkRule.ts
import type Delimiter from "../types/Delimiter";
import type InlineParserState from "../types/InlineParserState";
import type InlineRule from "../types/InlineRule";
import type LinkReference from "../types/LinkReference";
import type MarkdownNode from "../types/MarkdownNode";
import {
	BACKSLASH_CODE,
	BRACKET_OPEN_CODE,
	BRACKET_CLOSE_CODE,
	PAREN_OPEN_CODE,
	EXCLAMATION_CODE,
} from "../utils/charCodes";
import newText from "../utils/newText";
import normalizeLabel from "../utils/normalizeLabel";
import parseLinkInline from "../utils/parseLinkInline";

const rule: InlineRule = {
	name: "link",
	test: testLink,
	precedence: 15,
};
export default rule;

function testLink(state: InlineParserState, parent: MarkdownNode): boolean {
	if (!state.isEscaped) {
		let charCode = state.src.charCodeAt(state.i);

		if (charCode === BRACKET_OPEN_CODE) {
			return testLinkOpen(state, parent);
		}

		if (charCode === EXCLAMATION_CODE && state.src.charCodeAt(state.i + 1) === BRACKET_OPEN_CODE) {
			return testImageOpen(state, parent);
		}

		if (charCode === BRACKET_CLOSE_CODE) {
			return testLinkClose(state, parent);
		}
	}

	return false;
}

function testLinkOpen(state: InlineParserState, parent: MarkdownNode) {
	let start = state.i;
	let markup = "[";

	// Add a new text node which may turn into a link
	let text = newText(state.parentIndex + start, state.line, markup, 0);
	parent.children!.push(text);

	state.i++;
	state.delimiters.push({ markup, start, length: 1, precedence: rule.precedence });

	return true;
}

function testImageOpen(state: InlineParserState, parent: MarkdownNode) {
	let start = state.i;
	let markup = "![";

	// Add a new text node which may turn into an image
	let text = newText(state.parentIndex + start, state.line, markup, 0);
	parent.children!.push(text);

	state.i += markup.length;
	state.delimiters.push({ markup, start, length: 1, precedence: rule.precedence });

	return true;
}

function testLinkClose(state: InlineParserState, parent: MarkdownNode) {
	// Loop backwards through delimiters to find a matching one that does
	// not take precedence
	let startDelimiter: Delimiter | undefined;
	let i = state.delimiters.length;
	while (i--) {
		let prevDelimiter = state.delimiters[i];
		if (!prevDelimiter.handled) {
			if (prevDelimiter.markup === "[" || prevDelimiter.markup === "![") {
				startDelimiter = prevDelimiter;
				break;
			} else if ((prevDelimiter.precedence ?? 0) <= rule.precedence!) {
				// Same or lower precedence delimiters can be skipped over
				continue;
			} else {
				// Higher precedence delimiters block
				break;
			}
		}
	}

	if (startDelimiter !== undefined) {
		// Convert the text node into a link node with a new text child
		// followed by the other children of the parent (if any)
		let i = parent.children!.length;
		while (i--) {
			let lastNode = parent.children![i];
			if (lastNode.index === state.parentIndex + startDelimiter.start) {
				let start = state.i + 1;
				let label = state.src.substring(
					startDelimiter.start + startDelimiter.markup.length,
					state.i,
				);

				// "The link text may contain balanced brackets, but not
				// unbalanced ones, unless they are escaped"
				let level = 0;
				for (let i = 0; i < label.length; i++) {
					if (label.charCodeAt(i) === BACKSLASH_CODE) {
						i++;
					} else if (label.charCodeAt(i) === BRACKET_OPEN_CODE) {
						level++;
					} else if (label.charCodeAt(i) === BRACKET_CLOSE_CODE) {
						level--;
					}
				}
				if (level != 0) {
					return false;
				}

				let isLink = startDelimiter.markup === "[";

				let hasInfo = state.src.charCodeAt(state.i + 1) === PAREN_OPEN_CODE;
				let hasRef = state.src.charCodeAt(state.i + 1) === BRACKET_OPEN_CODE;

				// "Full and compact references take precedence over shortcut references"
				// "Inline links also take precedence"
				let link: LinkReference | undefined;
				if (hasInfo) {
					start++;
					link = parseLinkInline(state, start, ")");
					if (link !== undefined) {
						//state.i = ??
					}
				} else if (hasRef) {
					start++;
					for (let i = start; i < state.src.length; i++) {
						if (state.src.charCodeAt(i) === BRACKET_CLOSE_CODE) {
							// Lookup using the text between [], or if there
							// is no text, use the label
							label = i - start > 0 ? state.src.substring(start, i) : label;
							label = normalizeLabel(label);
							link = state.refs[label];
							if (link !== undefined) {
								state.i = i + 1;
							}
							break;
						}
					}
				}

				if (link === undefined) {
					label = normalizeLabel(label);
					link = state.refs[label];
					if (link !== undefined) {
						state.i++;
					}
				}

				if (link !== undefined) {
					let content = lastNode.content.slice(startDelimiter.markup.length);
					let text = newText(lastNode.index, lastNode.line, content, 0);
					text.length = text.content.length;

					lastNode.type = isLink ? "link" : "image";
					lastNode.info = link.url;
					lastNode.title = link.title;
					lastNode.children = [text, ...parent.children!.splice(i + 1)];
					lastNode.length = state.parentIndex + state.i - lastNode.index;

					// "[L]inks may not contain other links, at any level of nesting"
					if (isLink) {
						// Remove all the opening delimiters so they won't be picked up in future
						let d = state.delimiters.length;
						while (d--) {
							let prevDelimiter = state.delimiters[d];
							if (prevDelimiter.markup === "[" || prevDelimiter.markup === "]") {
								prevDelimiter.handled = true;
							}
						}
					}

					startDelimiter.handled = true;
					return true;
				}

				// TODO: If it's not a link, go back and close delimiters that
				// weren't closed between start and end

				startDelimiter.handled = true;
				break;
			}
		}
	}

	return false;
}
